pkg/cli/image/manifest/source.go

```diff
@@ -0,0 +1,141 @@
+		}
+		if len(icspData) == 0 {
+			return fmt.Errorf("no data found in ImageContentSourceFile %s", o.ImageContentSourcePolicyFile)
+		}
+		icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+		if err != nil {
+			return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		var icsp *operatorv1alpha1.ImageContentSourcePolicy
+		var ok bool
+		if icsp, ok = icspObj.(*operatorv1alpha1.ImageContentSourcePolicy); !ok {
+			return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", o.ImageContentSourcePolicyFile)
+		}
+		o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, *icsp)
+	}
+	o.ICSPClientFn = func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error) {
+		// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+		if len(o.ImageContentSourcePolicyFile) != 0 {
+			return nil, nil
+		}
+		restConfig, err := f.ToRESTConfig()
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		icspClient, err := operatorv1alpha1client.NewForConfig(restConfig)
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		return icspClient.ImageContentSourcePolicies(), nil
+	}
+	return nil
+}
+
+// GetICSPs will lookup ICSPs from cluster. Since it's not a hard requirement to find ICSPs from cluster, GetICSPs logs errors rather than returning errors.
+func (o *ImageSourceOptions) GetICSPs(icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface) {
+	icsps, err := icspClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
+		klog.V(4).Infof("did not access any ImageContentSourcePolicies in cluster: %v", err)
+	}
+	if len(icsps.Items) == 0 {
+		klog.V(4).Info("no ImageContentSourcePolicies found in cluster")
+	}
+	o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, icsps.Items...)
+}
+
+// GetImageSourcePrefixes returns list of 'registry/repository/name' options to try
+// 'registry/repository/name' determined from examining ImageContentSourcePolicy objects.
+// If a valid prefix is obtained from ICSP, return true
+func GetImageSourcePrefixes(icsps []operatorv1alpha1.ImageContentSourcePolicy, image string) (bool, []string, error) {
+	var imageSources []string
+	for _, icsp := range icsps {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		imageRef, err := reference.Parse(image)
```

Comment:
You don't need to re-parse it on every iteration.

Reply:
fixed, updated